modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1403 @@
+    t.stop();
+    if retval & EV_TIMEOUT != 0 {
+      throw SystemError.fromSyserr(ETIMEDOUT, "send timed out");
+    }
+    var elapsedTime = t.elapsed(TimeUnits.microseconds):c_long;
+    err_out = sys_sendto(this.socketFd, data.c_str():c_void_ptr, data.size:c_long, 0, address._addressStorage, length);
+    if err_out != 0 {
+      if err_out != EAGAIN && err_out != EWOULDBLOCK {
+        throw SystemError.fromSyserr(err_out, "send failed");
+      }
+      if timeout.tv_sec == -1 {
+        var totalTimeout = timeout.tv_sec*1000000 + timeout.tv_usec;
+        if totalTimeout >= t.elapsed(TimeUnits.microseconds) {
+          const remainingMicroSeconds = ((totalTimeout - elapsedTime)%1000000);
+          const remainingSeconds = ((totalTimeout - elapsedTime)/1000000);
+          timeout.tv_sec = remainingSeconds;
+          timeout.tv_usec = remainingMicroSeconds;
+        }
+        throw SystemError.fromSyserr(ETIMEDOUT, "send timed out");
+      }
+    }
+  }
+  return length;
+}
+
+inline operator !=(const ref lhs: udpSocket,const ref rhs: udpSocket) {
+  return lhs.socketFd != rhs.socketFd;
+}
+
+inline operator ==(const ref lhs: udpSocket,const ref rhs: udpSocket) {
+  return lhs.socketFd == rhs.socketFd;
+}
+
+proc udpSocket.writeThis(f) throws {
+  f.write("(","addr:",this.addr,",fd:",this.socketFd);
+}
+
+pragma "no doc"
+proc setSockOpt(socketFd: fd_t, level: c_int, optname: c_int, ref value: c_int) throws {
+  var optlen = sizeof(value):socklen_t;
+  var ptroptval = c_ptrTo(value);
+  var err_out = sys_setsockopt(socketFd, level, optname, ptroptval:c_void_ptr, optlen);
+  if err_out != 0 {
+    throw SystemError.fromSyserr(err_out, "Failed to set socket option");
+  }
+}
+
+/*
+  Set the value of the given socket option (see `setsockopt(2) </https://pubs.opengroup.org/onlinepubs/9699919799/functions/setsockopt.html#>`_)
+  on provided :type:`tcpConn`. The needed symbolic constants (SO_* etc.)
+  are defined in the :mod:`Sys`.
+
+  .. code-block:: Chapel
+
+    setSockOpt(socket, IPPROTO_TCP, TCP_QUICKACK, 1:c_int);
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+  :arg level: protocol level
+  :type level: `int(32)`
+  :arg optname: option to set.
+  :type optname: `int(32)`
+  :arg value: value to set on option
+  :type value: `int(32)`
+  :throws SystemError: Upon incompatible arguments
+                        and socket.
+*/
+proc setSockOpt(ref socket: tcpConn, level: c_int, optname: c_int, value: c_int) throws {
+  var socketFd = socket.socketFd;
+  setSockOpt(socketFd, level, optname, value);
+}
+
+/*
+  See above
+*/
+proc setSockOpt(ref socket: tcpListener, level: c_int, optname: c_int, value: c_int) throws {
+  var socketFd = socket.socketFd;
+  setSockOpt(socketFd, level, optname, value);
+}
+
+/*
+  See above
+*/
+proc setSockOpt(ref socket: udpSocket, level: c_int, optname: c_int, value: c_int) throws {
+  var socketFd = socket.socketFd;
+  setSockOpt(socketFd, level, optname, value);
+}
```

Comment:
For all these 3 overloads, I think a better way is to `no doc` the 2nd and 3rd ones, make the 1st one's argument generic, add a `where` clause to it to type-check, change its documentation to say that it accepts 3 different types.

Reply:
I see that this overloads go on and on a bit more. I think we should think about reducing this number. To be clear, I talk about the documentation, I don't mind about the number of overloads in code.